examples/controlplane/kustomizeversions.yaml

```diff
@@ -0,0 +1,226 @@
+kind: KubeadmConfig
+metadata:
+  name: ${CLUSTER_NAME}-controlplane-0
+spec:
+  clusterConfiguration:
+    kubernetesVersion: "ci/latest"
+  preKubeadmCommands:
+  - bash -c /tmp/kubeadm-bootstrap.sh
+  files:
+  - path: /tmp/kubeadm-bootstrap.sh
+    owner: "root:root"
+    permissions: "0744"
+    content: |
+      #!/bin/bash
+
+      set -o nounset
+      set -o pipefail
+      set -o errexit
+      set -x
+
+      [[ $(id -u) != 0 ]] && SUDO="sudo" || SUDO=""
+
+      if ! gsutil version > /dev/null; then
+      curl -sSL https://sdk.cloud.google.com > /tmp/gcl && bash /tmp/gcl --install-dir=~/gcloud --disable-prompts
+      export PATH=$PATH:~/gcloud/google-cloud-sdk/bin
+      fi
+      gsutil version
+
+      # This test installs debian packages that are a result of the CI and release builds.
+      # It runs '... --version' commands to verify that the binaries are correctly installed
+      # and finally uninstalls the packages.
+      # For the release packages it tests all versions in the support skew.
+
+      LINE_SEPARATOR="*************************************************"
+      echo "$LINE_SEPARATOR"
+
+      declare -a PACKAGES_TO_TEST=("kubectl" "kubelet" "kubeadm")
+      PACKAGE_EXT="deb"
+      declare -a CONTAINERS_TO_TEST=("kube-apiserver" "kube-controller-manager" "kube-proxy" "kube-scheduler")
+      CONTAINER_EXT="tar"
+
+      CI_DIR=/tmp/k8s-ci
+      mkdir -p $CI_DIR
+      CI_VERSION=$(curl -sSL https://dl.k8s.io/ci/latest-green.txt)
+      CI_URL="gs://kubernetes-release-dev/ci/$CI_VERSION-bazel/bin/linux/amd64"
+
+      echo "* testing CI version $CI_VERSION"
+
+      for CI_PACKAGE in "${PACKAGES_TO_TEST[@]}"; do
+      echo "* downloading package: $CI_URL/$CI_PACKAGE.$PACKAGE_EXT"
+      gsutil cp "$CI_URL/$CI_PACKAGE.$PACKAGE_EXT" "$CI_DIR/$CI_PACKAGE.$PACKAGE_EXT"
+      ${SUDO} dpkg -i "$CI_DIR/$CI_PACKAGE.$PACKAGE_EXT" || echo "* ignoring expected 'dpkg -i' result"
+      done
+
+      CONTAINERD_VERSION="${CONTAINERD_VERSION:-1.3.0}"
+      wget --tries 10 https://storage.googleapis.com/cri-containerd-release/cri-containerd-${CONTAINERD_VERSION}.linux-amd64.tar.gz
+      tar --no-overwrite-dir -C $CI_DIR -xzf cri-containerd-${CONTAINERD_VERSION}.linux-amd64.tar.gz
+      export PATH=$CI_DIR/usr/local/bin:$PATH
+
+      for CI_CONTAINER in "${CONTAINERS_TO_TEST[@]}"; do
+      echo "* downloading package: $CI_URL/$CI_CONTAINER.$CONTAINER_EXT"
+      gsutil cp "$CI_URL/$CI_CONTAINER.$CONTAINER_EXT" "$CI_DIR/$CI_CONTAINER.$CONTAINER_EXT"
+      ${SUDO} ctr -n k8s.io images import "$CI_DIR/$CI_CONTAINER.$CONTAINER_EXT" || echo "* ignoring expected 'ctr images import' result"
+      ${SUDO} ctr -n k8s.io images tag k8s.gcr.io/$CI_CONTAINER-amd64:"${CI_VERSION//+/_}" k8s.gcr.io/$CI_CONTAINER:"${CI_VERSION//+/_}"
+      ${SUDO} ctr -n k8s.io images tag k8s.gcr.io/$CI_CONTAINER-amd64:"${CI_VERSION//+/_}" gcr.io/kubernetes-ci-images/$CI_CONTAINER:"${CI_VERSION//+/_}"
```

Comment:
ok so what we do in kinder tests is slightly different:
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-kustomize-master/1183740976779235328/build-log.txt

- we don't use `ci/latest` in "ClusterConfiguration -> kubernetesVersion".
- instead we resolve the `ci/latest` label to a version - e.g. `1.17.0-alpha.1.415_81a8b9804a3bf3`
- use this version for `kubernetesVersion`. [1]
- download image tarbals from
`https://storage.googleapis.com/kubernetes-release-dev/ci/VERSION/bin/linux/amd64/`
- create images from the tarbals while dropping the `-amd64` suffix.
- make sure the images are tagged as, e.g.:
`k8s.gcr.io/kube-apiserver:v1.17.0-alpha.1.415_81a8b9804a3bf3`

this tricks kubeadm into thinking that this is not a CI version but rather that `1.17.0-alpha.1.415_81a8b9804a3bf3` is a release version and results in using `k8s.gcr.io/` instead of the CI image bucket.

i realize that [1] might be a bit of a problem as `kubernetesVersion` is part of the same manifest where we inline the script to download artifacts, so the inlined script cannot apply that resolved CI version and kubeadm will try to use the CI bucket URLs for the images.

-------

testing with `ci/latest` locally:
- `ci/latest` resolved to `1.17.0-alpha.1.418_f7992c7a214cc9`
- kubeadm tries to download these images:
```
I1014 18:04:16.235147    3239 checks.go:845] pulling gcr.io/kubernetes-ci-images/kube-apiserver:v1.17.0-alpha.1.418_f7992c7a214cc9
I1014 18:04:17.834539    3239 checks.go:845] pulling gcr.io/kubernetes-ci-images/kube-controller-manager:v1.17.0-alpha.1.418_f7992c7a214cc9
I1014 18:04:19.311796    3239 checks.go:845] pulling gcr.io/kubernetes-ci-images/kube-scheduler:v1.17.0-alpha.1.418_f7992c7a214cc9
I1014 18:04:22.649278    3239 checks.go:845] pulling gcr.io/kubernetes-ci-images/kube-proxy:v1.17.0-alpha.1.418_f7992c7a214cc9
I1014 18:04:24.313702    3239 checks.go:845] pulling k8s.gcr.io/pause:3.1
I1014 18:04:27.687015    3239 checks.go:845] pulling k8s.gcr.io/etcd:3.3.15-0
I1014 18:05:06.829218    3239 checks.go:845] pulling k8s.gcr.io/coredns:1.6.2
```
- checking:
https://console.cloud.google.com/gcr/images/kubernetes-ci-images/GLOBAL/kube-apiserver?gcrImageListsize=30
seems that the latest api server is at `v1.17.0-alpha.1.411_376b0ae20e7c00`
- so i guess this is a related problem to https://github.com/kubernetes/release/issues/897
as `ci/latest` points to image tags that are still missing.

probably passing `ci/latest` to kubeadm directly is not a good idea for our tests in CAPG.

Reply:
i am fetching the contents of latest.txt and using that instead.

thanks,
Dims

Reply:
we chatted with tpepper about latest.txt today.
looks like the problems with missing artifacts is now gone for now (stabilized on the 12th):
https://storage.googleapis.com/k8s-gubernator/triage/index.html
but i suspect just the GCS bucket networking improved or something.